blur/rand.py

```diff
@@ -413,42 +409,53 @@ def weighted_shuffle(weights):
-        available_indexes = [i for i in range(len(weights))
-                             if i not in taken_positions]
-        return min((available_index for available_index in available_indexes),
-                   key=lambda index: abs(move_index - index))
-
-    while working_list:
-        # Pick which item to move
-        move_index = weighted_choice(
-            [(index, weight[2]) for index, weight in enumerate(working_list)])
-        move_item = working_list[move_index][0]
-        # Find the index where the item will be placed
-        if isinstance(working_list[move_index][1], str):
-            if working_list[move_index][1] == 'STAY':
-                # Place in the index closest to where the item appears already
-                target_position = closest_available(move_index)
+
+    def closest_available(requested_index, available_indices):
+        """Finds the closest available index to the requested index.
+        If there are two "closest" indexes, a random one will be returned."""
+        closest = None
+        min_distance = float("inf")
+        for index in available_indices:
+            distance = abs(requested_index - index)
+            if distance <= min_distance:
+                min_distance = distance
+                closest = index
+                # instantiate closest or overwrite it 50% of the time if exists
+                if closest is None or random.getrandbits(1):
+                    closest = index
+        return closest
+
+    def requested_index(i, requested_pos, length):
+        """Calculates an absolute index for a requested position, where
+        a position is either a percentage or the string 'STAY'."""
+        if isinstance(requested_pos, str):
+            if requested_pos == 'STAY':
+                return i
             else:
-                raise ValueError
+                msg = "'{0}' is not a valid str argument "
+                "for a requested position.".format(requested_pos)
+                raise ValueError(msg)
         else:
-            # Place in the index closest to working_list[][1] percent along
-            requested_index = int((working_list[move_index][1] / 100) *
-                                  move_index)
-            target_position = closest_available(requested_index)
-        shuffle_positions.append((move_item, target_position))
-        # Remove the item weight from working_list
-        working_list.pop(move_index)
-
-    # Construct the shuffled list and return it
+            return int(round(requested_pos / 100 * (length - 1)))
+
+    weighted_absolute_positions = [
+        (value, requested_index(i, requested_pos, len(weights)), weight)
+        for i, (value, requested_pos, weight)
+        in enumerate(weights)
+        ]
+
+    # shuffles list to remove bias when sorting
+    random.shuffle(weighted_absolute_positions)
+    # sorts list in descending weights
+    weighted_absolute_positions.sort(key=lambda (_, __, w): w, reverse=True)

```

Comment:
This is probably bad python, but _ for non-needed values in pattern matches is pretty ingrained in me. Python doesn't allow duplicate `_` params.